### DISCUSSION TITLE
Can I regenerate a ISR page based on user ip address?

### DISCUSSION THREAD

**Discussion by aomini** [vercel/next.js#38500] [Help] [answered]
I want to be able to pass the ip address of the user who triggered the ISR. But also the static page should be cached based on his location.

**Comment by icyJoseph** [comment, marked as answer]
I guess you could combine `middleware`, the routing based on the file system, and `getStaticPaths` and `getStaticProps`:

1. In the middleware rewrite requests to `/_site/[ip]/content`,
2. Create a folder `/_site` in `pages`, and inside another folder named `[ip]`, in here you place all of your site pages
3. The middleware should reject requests if the path starts with `/_site`, to prevent direct access
4. In any page of `[ip]`, and here's the catch with this approach, you need to implement `getStaticPaths` and `getStaticProps`
5. The `getStaticPaths` should result in an empty list, and fallback true.
6. In `getStaticProps` you generate content using the IP
7. You need to account for a fallback page via `useRouter().isFallback`, for pages for IP's that have not yet been generated
8. You also probably need a custom 404 page

> One key assumption is that you'll be able to pinpoint the user IP in the middleware. 

> It might not be easy to test locally

#### middleware

```ts
import { NextResponse, type NextRequest } from "next/server";

const PUBLIC_FILE = /\.(.*)$/;

export function middleware(req: NextRequest) {
  const { pathname } = req.nextUrl;
  const userIP = req.ip;

  console.log("Middleware", { pathname, userIP });

  if (
    pathname.startsWith("/_next") || // exclude Next.js internals
    pathname.startsWith("/api") || //  exclude all API routes
    pathname.startsWith("/static") || // exclude static files
    pathname.includes(".") || // exclude all files in the public folder
    PUBLIC_FILE.test(pathname)
  )
    return NextResponse.next();

  if (pathname.startsWith("/_site")) {
    req.nextUrl.pathname = `/_site/404`;
    return NextResponse.rewrite(req.nextUrl);
  }

  // perhaps you don't want to do this to all paths...
  if (pathname === "/") {
    req.nextUrl.pathname = `/_site/${userIP || ""}`;

    return NextResponse.rewrite(req.nextUrl);
  }

  return NextResponse.next();
}
```

### pages

```tsx
// pages/_site/[ip]/index.tsx
import type { GetStaticPaths, GetStaticProps, NextPage } from "next";
import { useRouter } from "next/router";

const IPBased: NextPage<{ ip: string }> = ({ ip }) => {
  const router = useRouter();

  if (router.isFallback) return <div>Preparing your page...</div>;

  return <div>Hello, your IP is: {ip}</div>;
};

export const getStaticPaths: GetStaticPaths<{ ip: string }> = () => {
  return { paths: [], fallback: true };
};

export const getStaticProps: GetStaticProps = (ctx) => {
  const ip = ctx.params?.ip;

  if (!ip) return { notFound: true };
  if (Array.isArray(ip)) return { notFound: true };

  return { props: { ip } };
};

export default IPBased;
```

```
pages/
├── _app.tsx
├── _document.tsx
├── _site
│   ├── 404.tsx
│   └── [ip]
│       └── index.tsx
├── api
│   └── hello.ts
├── blog.tsx
└── index.tsx
```

#### 404 page for `/_site`

Because middleware doesn't support body response anymore, we could just send unwanted requests to this page, which will just return the regular 404 page.

```tsx
// pages/_site/404.tsx
const Site404 = () => {
  return <div></div>;
};

export const getServerSideProps = () => {
  return { notFound: true };
};

export default Site404;
```

**Reply by aomini** [reply]
Thanks for the approach it has given me some idea. 
However, I cannot add the IP to URL. I want to generate `_site/[slug]` page. If I generate a page called `_site/product-1` with a ip from US then another user from another country will get the same page generated from the US IP. Is there any way I can make use of ISR and still serve my static pages depending upon ip address.

```
pages/
├── _app.tsx
├── _document.tsx
├── _site
│   ├── 404.tsx
│   └── [slug].tsx
├── api
│   └── hello.ts
├── blog.tsx
└── index.tsx
```

**Reply by icyJoseph** [reply]
> However, I cannot add the IP to URL. 

The rewrites approach doesn't add it to the visible URL, only to the URL handled by the next routing system. The user still sees `/product-id`, but the routing internals see `/_site/ip/product-id`.

>  If I generate a page called _site/product-1 with a ip from US then another user from another country will get the same page generated from the US IP. Is there any way I can make use of ISR and still serve my static pages depending upon ip address.

For this case, you'd need to handle this in another layer I'd say. 

In other words, as far as the Next.js server can see, the requests are the same. You could, for example, use lambda at edge from AWS, and if there's no cache hit for the requested page and IP pair, then let the request go to Next.js, and cache Next.js's response for that route and that IP.

**Reply by aomini** [reply]
Thank you very much. These clear things up.
